sw/airborne/modules/ins/ins_ext_pose.c

```diff
@@ -0,0 +1,1246 @@
+	float x1=cos(X[8]);
+	float x2=sin(X[8]);
+	float x3=cos(X[6]);
+	float x4=x2*x3;
+	float x5=sin(X[7]);
+	float x6=sin(X[6]);
+	float x7=x1*x6;
+	float x8=x2*x6;
+	float x9=x1*x3;
+	float x10=tan(X[7]);
+	float x11=1.0/x0;
+	out[0][0]=0;
+	out[0][1]=0;
+	out[0][2]=0;
+	out[0][3]=0;
+	out[0][4]=0;
+	out[0][5]=0;
+	out[1][0]=0;
+	out[1][1]=0;
+	out[1][2]=0;
+	out[1][3]=0;
+	out[1][4]=0;
+	out[1][5]=0;
+	out[2][0]=0;
+	out[2][1]=0;
+	out[2][2]=0;
+	out[2][3]=0;
+	out[2][4]=0;
+	out[2][5]=0;
+	out[3][0]=-x0*x1;
+	out[3][1]=x4 - x5*x7;
+	out[3][2]=-x5*x9 - x8;
+	out[3][3]=0;
+	out[3][4]=0;
+	out[3][5]=0;
+	out[4][0]=-x0*x2;
+	out[4][1]=-x5*x8 - x9;
+	out[4][2]=-x4*x5 + x7;
+	out[4][3]=0;
+	out[4][4]=0;
+	out[4][5]=0;
+	out[5][0]=x5;
+	out[5][1]=-x0*x6;
+	out[5][2]=-x0*x3;
+	out[5][3]=0;
+	out[5][4]=0;
+	out[5][5]=0;
+	out[6][0]=0;
+	out[6][1]=0;
+	out[6][2]=0;
+	out[6][3]=-1;
+	out[6][4]=-x10*x6;
+	out[6][5]=-x10*x3;
+	out[7][0]=0;
+	out[7][1]=0;
+	out[7][2]=0;
+	out[7][3]=0;
+	out[7][4]=-x3;
+	out[7][5]=x6;
+	out[8][0]=0;
+	out[8][1]=0;
+	out[8][2]=0;
+	out[8][3]=0;
+	out[8][4]=-x11*x6;
+	out[8][5]=-x11*x3;
+	out[9][0]=0;
+	out[9][1]=0;
+	out[9][2]=0;
+	out[9][3]=0;
+	out[9][4]=0;
+	out[9][5]=0;
+	out[10][0]=0;
+	out[10][1]=0;
+	out[10][2]=0;
+	out[10][3]=0;
+	out[10][4]=0;
+	out[10][5]=0;
+	out[11][0]=0;
+	out[11][1]=0;
+	out[11][2]=0;
+	out[11][3]=0;
+	out[11][4]=0;
+	out[11][5]=0;
+	out[12][0]=0;
+	out[12][1]=0;
+	out[12][2]=0;
+	out[12][3]=0;
+	out[12][4]=0;
+	out[12][5]=0;
+	out[13][0]=0;
+	out[13][1]=0;
+	out[13][2]=0;
+	out[13][3]=0;
+	out[13][4]=0;
+	out[13][5]=0;
+	out[14][0]=0;
+	out[14][1]=0;
+	out[14][2]=0;
+	out[14][3]=0;
+	out[14][4]=0;
+	out[14][5]=0;
+}
+
+
+
+static inline void ekf_f_rk4(const float X[EKF_NUM_STATES], const float U[EKF_NUM_INPUTS], const float dt, float out[EKF_NUM_STATES])
+{
+	float k1[EKF_NUM_STATES];
+	float k2[EKF_NUM_STATES];
+	float k3[EKF_NUM_STATES];
+	float k4[EKF_NUM_STATES];
+
+	float Xtmp[EKF_NUM_STATES];
+
+	// k1   = f(X,U)
+	ekf_f(X,U,k1);
+	
+	// Xtmp = X+dt*k1/2
+	float_vect_smul(Xtmp, k1, dt/2, EKF_NUM_STATES);
+	float_vect_add(Xtmp, X, EKF_NUM_STATES);
+	
+	// k2   = f(Xtmp,U)
+	ekf_f(Xtmp,U,k2);
+	
+	// Xtmp = X+dt*k2/2
+	float_vect_smul(Xtmp, k2, dt/2, EKF_NUM_STATES);
+	float_vect_add(Xtmp, X, EKF_NUM_STATES);
+
+	// k3   = f(Xtmp,U)
+	ekf_f(Xtmp,U,k3);
+
+	// Xtmp = X+dt*k3
+	float_vect_smul(Xtmp, k3, dt, EKF_NUM_STATES);
+	float_vect_add(Xtmp, X, EKF_NUM_STATES);
+
+	// k4   = f(Xtmp,U)
+	ekf_f(Xtmp,U,k4);	
+	
+	// out = k2+k3
+	float_vect_sum(out, k2, k3, EKF_NUM_STATES);
+	// out *= 2
+	float_vect_scale(out, 2, EKF_NUM_STATES);
+	// out += k1
+	float_vect_add(out, k1, EKF_NUM_STATES);
+	// out += k4
+	float_vect_add(out, k4, EKF_NUM_STATES);
+	// out *= dt/6
+	float_vect_scale(out, dt/6, EKF_NUM_STATES);
+	// out += X
+	float_vect_add(out, X, EKF_NUM_STATES);	
+}
+
+
+static inline void ekf_step(const float U[EKF_NUM_INPUTS], const float Z[EKF_NUM_OUTPUTS], const float dt)
```

Comment:
would it be possible to make a generic library like https://github.com/paparazzi/paparazzi/blob/master/sw/airborne/filters/linear_kalman_filter.h ?

Reply:
Of course it means that you would have to pass functions for the model and the jacobian, but that is for an other day